### PR TITLE
Python wrapper

### PR DESCRIPTION
**Pull request by Thomasb81**
Find here an attempt to fix  https://github.com/chipsalliance/Surelog/issues/2635
I try to follow guidance that have been provide.

- The wrapping of VPI method only has a limited usage since it does not allow to change the uhdm db.
- Some VPI method implementation were missing, so dummy (empty) implementation has been provided.
- Some VPI method have not been wrap because they are variadic function and wrapping with swig was not obvious...
- swig 4.0 limitation : UHDM C++ 17 code is not fully supported. `#ifdef` / `#ifndef` SWIG have been introduce possibly in generated code. std:string_view is not yet supported and has been work-arround by creating new method using std::string as argument. Not clear to me if there is more efficient and how to implement it.
- swig wrapper generation still produce warnings, can't figure out the consequence now:
```
/home/tom/prog/git/UHDM/dbuild/generated/uhdm/vpi_user.h:656: Warning 312: Nested union not currently supported (ignored).
/home/tom/prog/git/UHDM/dbuild/generated/uhdm/vpi_user.h:674: Warning 312: Nested union not currently supported (ignored).
/home/tom/prog/git/UHDM/dbuild/generated/uhdm/sv_vpi_user.h:541: Warning 312: Nested union not currently supported (ignored).
/home/tom/prog/git/UHDM/dbuild/generated/uhdm/vpi_user.h:1027: Warning 462: Unable to set dimensionless array variable
```
- The build system has been amended to compile this shared library if UHDM is compile with  `BUILD_SHARED_LIBS` option set
- A single test is currently provide testing the proper import of uhdm python module. Additional tests should probably be provide and would help to demonstrate how to use the library.

### CODE REVIEW

**Review by alaindargelas — APPROVED** [review]

**Comment by alaindargelas** [comment]
@hs-apotell Can you review since this touches some code generation.

**Comment by Thomasb81** [comment]
I am not happy with the way I change `classes.py` to generate dot i file... There is some behavior of the script that I did not fully understand. But it is working. Code change can still be done later... The main point is : it did not break anything.

File: scripts/classes.py [hunk 1]
```diff
@@ -36,13 +36,23 @@ def _get_declarations(classname, type, vpi, card, real_type=''):
 
         if type == 'std::string':
             content.append(f'  {virtual}bool {Vpi_}(std::string_view data){final};')
+            content.append(f'#ifdef SWIG\n  {virtual}bool {Vpi_}(const std::string &data) {{ return {Vpi_}( std::string_view(data.c_str())); }}\n#endif')
```
[the remark this answers is not in the document]
Comment (Thomasb81) [reply scripts/classes.py:39]: Indeed it is.
Here is the typical error I face if I do not provide method with a `std::string` as argument but `std::string view`:
```
tom@UX430UA:~/prog/git/UHDM/dbuild/python$ ipython
Python 3.10.7 (main, Mar 10 2023, 10:47:39) [GCC 12.2.0]
Type 'copyright', 'credits' or 'license' for more information
IPython 7.31.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import uhdm

In [2]: s = uhdm.Serializer()

In [3]: design = s.MakeDesign()

In [4]: design.VpiName("design")
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-23bbe9f133bf> in <module>
----> 1 design.VpiName("design")

~/prog/git/UHDM/dbuild/python/uhdm.py in VpiName(self, *args)
  13752 
  13753     def VpiName(self, *args):
> 13754         return _py_uhdm.design_VpiName(self, *args)
  13755 
  13756     def VpiType(self):

TypeError: Wrong number or type of arguments for overloaded function 'design_VpiName'.
  Possible C/C++ prototypes are:
    design::VpiName(std::string_view)
    design::VpiName() const


In [5]: quit
```
The swig glue is unable to recognize the correct method to call. I would expect that in later version of swig c++17 will provide a better support of `std::string_view`

If you can suggest a different way to work-around this issue, I can give a trial.

File: scripts/serializer.py
[the remark this answers is not in the document]
Comment (Thomasb81) [reply scripts/serializer.py:1]: Right !

**Comment by Thomasb81** [comment]
> Looks like your approach is generating a parallel set of headers for use with swig. Any reason why the existing headers cannot be used without duplicating them? Even if required, wrapping certain functions with the SWIG macro would be preferred approach over duplicating the entire header.

Using the existing header was the initial approach with much more `#ifndef SIWG` / `#endif`.

Having separate file make it more clear of what is wrapped of what is not.
Also I though it is more easy to produce docstring or doxygen statement for automatic documentation generation.

But I agree this is high questionable.

**Comment by alaindargelas** [comment]
@Thomasb81 , in my original proposal of an implementation:
https://github.com/chipsalliance/Surelog/issues/2635#issuecomment-1077018391

I proposed we only expose the C API of UHDM (Read only) which is the VPI Standard interface. 
This is going to be 99% of the use cases.
No need to expose the C++ API for most use cases.
Can we limit this PR to do just that, or create another PR that solves that much easier problem first?

**Comment by Thomasb81** [comment]
> @Thomasb81 , in my original proposal of an implementation: [chipsalliance/Surelog#2635 (comment)](https://github.com/chipsalliance/Surelog/issues/2635#issuecomment-1077018391)
> 
> I proposed we only expose the C API of UHDM (Read only) which is the VPI Standard interface. This is going to be 99% of the use cases. No need to expose the C++ API for most use cases. Can we limit this PR to do just that, or create another PR that solves that much easier problem first?

It was my first intention. 

But then come the question : what can I provide as unit-test to ensure, next project contribution does not break this feature. The load of uhdm module is definitively not enough. We don't want to data-manage uhdm file, that can change if model change and coupling C++ for test vector generation with a python unit test was not simple enough. 

Then I though going a step further is achievable. I underestimate the complexity to handle the generation code from model and difficulty to wrap the some C++ code...

**Comment by alaindargelas** [comment]
You could have simply created a C++ function that creates "a model" from C++, invoke it from Python , then query the model from Python. 
You don't have to test many features, meaning you don't have to test UHDM creation or VPI use, that is tested plenty. Just make sure the most important vpi_iterate, get_get functions do work from Python.
I think you are now down a rabbit hole that is going be hard to see through.

**Comment by Thomasb81** [comment]
@alaindargelas @hs-apotell 
I simplified the current PR for vpi API and add a python demonstrating minimum iteration of uhdm db as you asked.
I expect this now meet your expectation.

Following python code should load an uhdm file:
```
import uhdm
s  = uhdm.Serializer()
data = s.Restore("a_uhdm_file.uhdm")
```

**Comment by yurivict** [comment]
Why is python binding dependent on ```BUILD_SHARED_LIBS``` [here](https://github.com/chipsalliance/UHDM/pull/895/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR391)?

Python binding should have nothing to do with whether the project is built with shared or with static libraries.

------------------

Also: why is Python bindings bundled with the UHDM project and is not made a separate project?

Multiple versions of Python can be installed concurrently, but UHDM can't be installed multiple times.
How can multiple versions of the bindings be installed in the situation of multiple Python versions if bindings is bundled with the UHDM project?

**Comment by Thomasb81** [comment]
> Why is python binding dependent on `BUILD_SHARED_LIBS` [here](https://github.com/chipsalliance/UHDM/pull/895/files#diff-1e7de1ae2d059d21e1dd75d5812d5a34b0222cef273b7c3a2af62eb747f9d20aR391)?

I am not aware of other method to create a python module from C or C++ code :  https://www.swig.org/Doc4.0/SWIGDocumentation.html#Introduction_nn8
Compiler complain if all the source code is not compile with `-fpic`. not sure we can mix `-fpic` at compilation and `-shared` at link stage

Unless you do the other way : integrate the python interpreter to your application. (What is done in Surelog, I think)

> Multiple versions of Python can be installed concurrently, but UHDM can't be installed multiple times.

Did I miss something ? I try to understand / follow cmake swig integration. As far as remember the installation is done in python version specific directories. At least on ubuntu.

Feel free to propose modification that will full fill your needs or what you think need to be done. I am not expert.

**Comment by maxslug** [comment]
One initial hurdle I found here was that Python needs to be compiled with `--enable-shared` to create `libpython3.x.so`.   This is fine for system versions of Python, but it turns out `pyenv` does not do this by default.
For future googlers, if you get the `-fpic` warning, it may be because of this. 

To compile shared w/ pyenv:
```
env PYTHON_CONFIGURE_OPTS="--enable-shared" pyenv install -verbose 3.9.11
```

**Pull request merged**